---
title: "Make ML trees"
output: html_document
date: "2025-08-27"
author: "Sofie Agerb√¶k"
---


Load alignment, trim data, subset data into partitions 
For each partition make a NJ tree, save these as newick files.

```{r, echo=FALSE}
library(seqinr)
library(ape)
library(ggplot2)
library(purrr)
library(dplyr)
```

## Load alignment
```{r}
genome_alignment <- read.alignment("HIV1_CON_2021_genome_DNA_subset.fasta", format = "fasta")
alignment_matrix <- as.DNAbin(genome_alignment)
```

## Trim data - remove columns with gaps in more than 50% of the sequences.  
```{r}
trimCols <- function(al, prop, codon = T){
  mat <- as.character(as.matrix(al))
  ntax <- nrow(mat)
  propthres <- 1-prop
  compliantSites <- apply(mat, 2, function(x){
    x <- as.character(x)
    compl <- (length(which(x %in% c("N", "n", "?", "-", "O", "o", "X", "x"))) / ntax) < propthres
    return(compl)
  })
  
  if(codon){
    codIDs <- rep(1:(length(compliantSites)/3), each = 3)
    codsToKeep <- rep(F, length(compliantSites))
    for(i in 1:max(codIDs)){
      if(all(compliantSites[which(codIDs == i)])) codsToKeep[which(codIDs == i)] <- T
    }
    al <- al[, as.logical(codsToKeep)]
    return(al)
  } else {
    al <- al[, as.logical(compliantSites)]
    return(al)
  }
}
```

```{r}
trimmed_alignment <- trimCols(alignment_matrix, 0.5, codon = FALSE)
alignment_matrix <- trimmed_alignment
```

## save trimmed alignment in fasta format
```{r}
sequences <- as.character(as.list(alignment_matrix))
sequences <- sapply(sequences, paste, collapse = "")
headers <- names(sequences)
```

Write trimmed alignment to file
```{r}
seq_along(sequences)

sequences[1]

fasta_lines <- unlist(lapply(names(sequences), function(label) {
  seq <- paste0(sequences[[label]], collapse = "") # Join the sequence
  c(paste0(">", label), seq) # Create a header and sequence entry
}))

output_file <- "2708_trimmed_alns_HIV_2021_subset.fasta"

writeLines(fasta_lines, output_file)
```

## Make Maximum likelihood trees with IQtree 

```{r}
Sys.setenv(PATH = paste("C:/Users/soage/iqtree-3.0.1-Windows/bin", Sys.getenv("PATH"), sep=";"))
```


Read trimmed alignment from fasta file
```{r}
genome_alignment <- read.alignment("2708_trimmed_alns_HIV_2021_subset.fasta", format = "fasta")
alignment_matrix <- as.DNAbin(genome_alignment)
```


Partitioning data into 18 segments (~500bp segment for each tree)

```{r}
alignment_length <- ncol(alignment_matrix)
num_sequences <- nrow(alignment_matrix)
window_size <- round(alignment_length / 18)

trees <- list()

for (start in seq(1, alignment_length, by = window_size)) {
  end <- start + window_size - 1
  print(start)
  
  if (end > alignment_length) {
    end <- alignment_length
  } #Make sure the last segment is not out of bounds
  print(end)
  #current window 
  segment <- alignment_matrix[, start:end]
  
  write.FASTA(segment, file = paste0("aln_segment_", start, "_to_", end, ".fasta"))
  file_path <- paste0("aln_segment_", start, "_to_", end)
 
  system(paste0("iqtree3 -s ", file_path, ".fasta -m MFP -B 1000 -T 4 -pre ", file_path))
  
  ml_tree <- read.tree(paste0(file_path, ".treefile"))
  trees[[paste0("Tree_", start, "_to_", end)]] <- ml_tree
}

multiPhylo <- do.call(c, trees)
write.tree(multiPhylo, file = "all_18_trimmed_ML_trees.newick")

```


We shift the windows by half a window length to create 17 trees for comparison of break points
```{r}
setwd("../17ML_trees")

half_window_size <- round(window_size/2)
trees17 <- list()

for (start in seq(half_window_size, alignment_length - half_window_size + 1, by = window_size)) {
  end <- start + window_size - 1
  print(start)
  print(end)
  
  #current window 
  segment <- alignment_matrix[, start:end]
  
  write.FASTA(segment, file = paste0("aln_segment_", start, "_to_", end, ".fasta"))
  file_path <- paste0("aln_segment_", start, "_to_", end)
 
  system(paste0("iqtree3 -s ", file_path, ".fasta -m MFP -B 1000 -T 4 -pre ", file_path))
  
  ml_tree <- read.tree(paste0(file_path, ".treefile"))
  trees17[[paste0("Tree_", start, "_to_", end)]] <- ml_tree
}

#Write file with all trees in multiphylo object
multiPhylo <- do.call(c, trees17)
write.tree(multiPhylo, file = "17_shifted_windows_trimmed_ML_trees.newick")

```

```{r}
for (start in seq(half_window_size, alignment_length - half_window_size + 1, by = window_size)) {
  end <- start + window_size - 1
  print(start)
  print(end)
  print(end - start)
  }
```



## Visualising branch lengths 

```{r}
trees_18_ML <- read.tree(file = "18ML_trees/all_18_trimmed_ML_trees.newick")

branch_data18 <- imap_dfr(trees_18_ML, ~{
  tibble(
    tree = .y,                      # tree index
    branch_length = .x$edge.length  # branch lengths
  )
})

p18 <- ggplot(branch_data18)+
  geom_histogram(aes(x=log(branch_length)), bins = 30, colour = "grey60")+
  labs(title = "18 original ML trees")+
  theme_minimal()+
  ylim(0,250)

p18

#Sanity check to make sure no tree has too different branch lengths
ggplot(branch_data, aes(x = factor(tree), y = log(branch_length), fill = factor(tree))) +
  geom_boxplot() +
  theme_minimal() +
  labs(
    x = "Tree",
    y = "Log Branch length",
    fill = "Tree",
    title = "Branch length distributions per tree"
  ) +
  theme(legend.position = "none") 

```

```{r}
trees_17_ML <- read.tree(file = "17ML_trees/17_shifted_windows_trimmed_ML_trees.newick")

branch_data17 <- imap_dfr(trees_17_ML, ~{
  tibble(
    tree = .y,                      # tree index
    branch_length = .x$edge.length  # branch lengths
  )
})

p17 <- ggplot(branch_data17)+
  geom_histogram(aes(x=log(branch_length)), bins = 30, colour = "grey60")+
  labs(title = "17 shifted ML trees")+
  theme_minimal()+
  ylim(0,250)
p17
```

```{r}
library(patchwork)
p18 + p17 
```


